### PR TITLE
Change compound assignment structure to support String concatenation

### PR DESCRIPTION
**Pull request by jdconrad**
This change modifies the compound assignment structure of the ir tree to support string concatenation in Java 8. This aligns with the changes already pulled into the backport (https://github.com/elastic/elasticsearch/pull/61755) up for review. The structure is changed so that when a StringBuilder is used for string concatenation it will be the first value pushed onto the stack prior loading the left-hand side of the compound assignment.

### CODE REVIEW

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/BinaryImplNode.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package org.elasticsearch.painless.ir;
+
+import org.elasticsearch.painless.ClassWriter;
+import org.elasticsearch.painless.MethodWriter;
+import org.elasticsearch.painless.phase.IRTreeVisitor;
+import org.elasticsearch.painless.symbol.WriteScope;
+
+public class BinaryImplNode extends BinaryNode {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ir/BinaryImplNode.java:27]: Brief comment on what this node is

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 2]
```diff
@@ -427,89 +426,63 @@ protected ExpressionNode injectCast(AExpression userExpressionNode, ScriptScope
...
+
+        // this load/store is a dot or brace load
+        if (irPrefixNode != null) {
+            // this load/store requires and index
+            if (irIndexNode != null) {
+                BinaryImplNode binaryImplNode = new BinaryImplNode();
+                binaryImplNode.setLocation(location);
+                binaryImplNode.setExpressionType(void.class);
+
+                if (isNullSafe) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:440]: add a line about the subtlety here.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 3]
```diff
@@ -862,7 +835,7 @@ public void visitAssignment(EAssignment userAssignmentNode, ScriptScope scriptSc
                 stringConcatenationNode.setLocation(irStoreNode.getLocation());
                 stringConcatenationNode.setExpressionType(String.class);
                 irCompoundNode = stringConcatenationNode;
-            // handles when the operation is mathematical
+                // handles when the operation is mathematical
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:838]: maybe put these after the control structure on the same line so intellij doesn't keep "intelligently" moving them.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java [hunk 4]
```diff
@@ -938,9 +911,11 @@ public void visitAssignment(EAssignment userAssignmentNode, ScriptScope scriptSc
                 irBinaryMathNode.setLeftNode(irLoadNode);
                 irBinaryMathNode.setRightNode(irValueNode);
             }
-        // handles a standard assignment
+
+            irAssignmentNode = irStoreNode;
+            // handles a standard assignment
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:916]: indented

**Review by stu-elastic — APPROVED** [review]
Nice change.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 2 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:440]: Done.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 3 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:838]: Fixed.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java (hunk 4 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultUserTreeToIRTreePhase.java:916]: Fixed.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review. Will commit once CI passes.

**Pull request merged**